Error when running with cDAQ modules

```python
Traceback (most recent call last):
  File "C:\Users\Wave\dev\makewaves\makewaves\waveio.py", line 133, in run
    daqmx.WriteAnalogF64(
  File "C:\Users\Wave\mambaforge\lib\site-packages\daqmx\core.py", line 345, in WriteAnalogF64
    ErrorHandling(rv, fatalerror)
  File "C:\Users\Wave\mambaforge\lib\site-packages\daqmx\core.py", line 617, in ErrorHandling
    raise RuntimeError(estring)
RuntimeError: DAC conversion attempted before data to be converted was available.
Decrease the output frequency to increase the period between DAC conversions, or reduce the size of your output buffer in order to write data more often. If you are using an external clock, check  your signal for the presence of noise or glitches.
make: *** [Makefile:5: run] Error 127
```

Should probably figure out how to get the callbacks working properly.

--- a/Makefile
+++ b/Makefile
@@ -2,7 +2,7 @@
 
 
 run:
-	@python -c "import makewaves; makewaves.main()"
+	@winpty python -c "import makewaves; makewaves.main()"
 
 
 ui:

--- a/makewaves/__init__.py
+++ b/makewaves/__init__.py
@@ -1,5 +1,5 @@
 
 from __future__ import division, print_function, absolute_import
 
-__version__ = "0.1.1"
+__version__ = "0.1.2"
 from .makewaves import *

--- a/makewaves/makewaves.py
+++ b/makewaves/makewaves.py
@@ -155,9 +155,13 @@ class MainWindow(QMainWindow):
         # Set default tab to regular waves
         self.ui.tabwidget.setCurrentIndex(0)
 
+        # Connect slots
+        self.connectslots()
+
         # Initialize slider values
         hmax = maxH[np.where(periods == np.round(1, decimals=2))[0]]
         self.ui.spinbox_wave_height.setMaximum(float(hmax))
+        self.ui.spinbox_wave_height.setValue(0.1)
         self.update_slider_height_value()
 
         # Initialize wavelength value
@@ -167,11 +171,12 @@ class MainWindow(QMainWindow):
             / wml.dispsolver(2 * np.pi / 1.0, water_depth, decimals=2)
         )
         self.ui.spinbox_wavelength.setValue(wl)
+        self.ui.spinbox_wave_period.setValue(1.0)
         self.update_slider_horiz_value()
+        self.update_slider_height_value()
 
         # Initialize plot settings
         self.initialize_plots()
-        self.connectslots()
         # Add dock widgets to right dock widget area and tabify them
         self.tabifyDockWidget(self.ui.dock_time_series, self.ui.dock_spectrum)
         self.on_rw_changed()
@@ -361,7 +366,7 @@ class MainWindow(QMainWindow):
         frac = (v - vmin) / (vmax - vmin)
         mi = self.ui.slider_horiz.minimum()
         mx = self.ui.slider_horiz.maximum()
-        self.ui.slider_height.setValue(int(mi + frac * (mx - mi)))
+        self.ui.slider_horiz.setValue(int(mi + frac * (mx - mi)))
 
     def on_regpars(self):
         """Decides which two parameters to use for regular waves"""

--- a/makewaves/makewaves.py
+++ b/makewaves/makewaves.py
@@ -440,15 +440,15 @@ class MainWindow(QMainWindow):
             self.calcthread.start()
 
     def on_start(self):
+        """Make waves."""
         if self.ui.action_start.isChecked() == True:
-            """Make waves"""
             self.ui.action_start.setText("Stop ")
             self.ui.action_start.setToolTip("Stop Generating Waves")
             self.ui.action_start.setIcon(QIcon(":icons/agt_stop.png"))
             self.ui.tabwidget.setEnabled(False)
             wavetype = self.ui.tabwidget.currentIndex()
             if wavetype == 0:
-                """Create regular waves"""
+                # Create regular waves
                 self.slabel.setText("Generating regular waves... ")
                 self.period = self.ui.spinbox_wave_period.value()
                 self.height = self.ui.spinbox_wave_height.value()
@@ -460,7 +460,7 @@ class MainWindow(QMainWindow):
                 self.wavegen.start()
 
             elif wavetype == 1:
-                """Create random waves."""
+                # Create random waves
                 rspec = self.ui.combobox_randwavetype.currentText()
                 self.slabel.setText("Generating " + rspec + " waves... ")
                 self.wavegen = WaveGen(
@@ -479,9 +479,8 @@ class MainWindow(QMainWindow):
                     ].value()
                 self.wavegen.start()
             self.timer.start(500)
-
         elif self.ui.action_start.isChecked() == False:
-            """Stop making waves"""
+            # Stop making waves
             self.slabel.setText("Stopping... ")
             self.wavegen.stop()
             self.wavegen.stopgen.finished.connect(self.on_wave_finished)

--- a/makewaves/waveio.py
+++ b/makewaves/waveio.py
@@ -138,7 +138,7 @@ class WaveGen(QThread):
             daqmx.Val_GroupByChannel,
             self.rampdown_ts,
         )
-        self.n_iterations += 1
+        self.n_iterations += 1.5
         self.sleep()
         # Write zeros to the buffer
         print("Writing a buffer of zeros")
@@ -150,7 +150,7 @@ class WaveGen(QThread):
             daqmx.Val_GroupByChannel,
             np.zeros(self.buffsize),
         )
-        self.n_iterations += 1
+        self.n_iterations += 1.5
         self.sleep()
         daqmx.StopTask(self.AOtaskHandle)
         daqmx.WaitUntilTaskDone(self.AOtaskHandle, timeout=10.0)
@@ -161,7 +161,7 @@ class WaveGen(QThread):
     def sleep(self):
         """Sleep between iterations.
 
-        Our goal is to always stay 100 ms ahead of the FIFO buffer being empty.
+        Our goal is to always stay ahead of the FIFO buffer being empty.
         """
         # Based on how long we've been running and how many iterations we've
         # done, compute a time to sleep
@@ -170,7 +170,7 @@ class WaveGen(QThread):
         iteration_period = self.period if self.wavetype == "Regular" else 1.0
         t_expected = iteration_period * self.n_iterations
         print("t_expected:", t_expected)
-        sleeptime = t_expected - t_total - 0.1
+        sleeptime = t_expected - t_total - 0.2
         if sleeptime > 0:
             print("Sleeping", sleeptime, "seconds")
             time.sleep(sleeptime)

--- a/makewaves/waveio.py
+++ b/makewaves/waveio.py
@@ -87,23 +87,24 @@ class WaveGen(QThread):
             rampup_ts,
         )
         daqmx.StartTask(self.AOtaskHandle)
-        self.sleep(time_taken=0.5)
-
-        # Set iteration variable to keep track of how many chunks of data
-        # have been written
-        i = 1
+        self.t_start = time.time()
+        self.n_iterations = 0
+        self.sleep()
 
         # Main running loop that writes data to DAQmx buffer
         while self.enable:
             t0 = time.time()
-            print("Writing main wave time series data iteration", i)
+            print(
+                "Writing main wave time series data iteration",
+                self.n_iterations,
+            )
             write_space_avail = daqmx.GetWriteSpaceAvail(self.AOtaskHandle)
             print("Write space available:", write_space_avail)
             if self.wavetype != "Regular":
-                if i >= 120:
-                    self.dataw = tsparts[i % 120, :]
+                if self.n_iterations >= 120:
+                    self.dataw = tsparts[self.n_iterations % 120, :]
                 else:
-                    self.dataw = tsparts[i, :]
+                    self.dataw = tsparts[self.n_iterations, :]
             daqmx.WriteAnalogF64(
                 self.AOtaskHandle,
                 self.buffsize,
@@ -112,16 +113,20 @@ class WaveGen(QThread):
                 daqmx.Val_GroupByChannel,
                 self.dataw,
             )
-            i += 1
-            self.sleep(time_taken=time.time() - t0)
+            self.n_iterations += 1
+            self.sleep()
 
         print("Output disabled; ramping down")
         # After disabled, initiate rampdown time series
         if self.wavetype != "Regular":
-            if i >= 120:
-                self.rampdown_ts = ramp_ts(tsparts[i % 120, :], "down")
+            if self.n_iterations >= 120:
+                self.rampdown_ts = ramp_ts(
+                    tsparts[self.n_iterations % 120, :], "down"
+                )
             else:
-                self.rampdown_ts = ramp_ts(tsparts[i, :], "down")
+                self.rampdown_ts = ramp_ts(
+                    tsparts[self.n_iterations, :], "down"
+                )
         else:
             self.rampdown_ts = ramp_ts(self.dataw, "down")
         print(f"Writing rampdown time series (len: {len(self.rampdown_ts)})")
@@ -133,6 +138,7 @@ class WaveGen(QThread):
             daqmx.Val_GroupByChannel,
             self.rampdown_ts,
         )
+        self.n_iterations += 1
         self.sleep()
         # Write zeros to the buffer
         print("Writing a buffer of zeros")
@@ -144,6 +150,7 @@ class WaveGen(QThread):
             daqmx.Val_GroupByChannel,
             np.zeros(self.buffsize),
         )
+        self.n_iterations += 1
         self.sleep()
         daqmx.StopTask(self.AOtaskHandle)
         daqmx.WaitUntilTaskDone(self.AOtaskHandle, timeout=10.0)
@@ -151,14 +158,19 @@ class WaveGen(QThread):
         daqmx.ClearTask(self.AOtaskHandle)
         self.rampeddown = True
 
-    def sleep(self, time_taken=0.0):
-        """Sleep between iterations."""
-        frac = 1.0
-        if self.wavetype == "Regular":
-            sleeptime = frac * self.period
-        else:
-            sleeptime = frac
-        sleeptime -= time_taken
+    def sleep(self):
+        """Sleep between iterations.
+
+        Our goal is to always stay 100 ms ahead of the FIFO buffer being empty.
+        """
+        # Based on how long we've been running and how many iterations we've
+        # done, compute a time to sleep
+        t_total = time.time() - self.t_start
+        print("t_total:", t_total)
+        iteration_period = self.period if self.wavetype == "Regular" else 1.0
+        t_expected = iteration_period * self.n_iterations
+        print("t_expected:", t_expected)
+        sleeptime = t_expected - t_total - 0.1
         if sleeptime > 0:
             print("Sleeping", sleeptime, "seconds")
             time.sleep(sleeptime)

--- a/makewaves/waveio.py
+++ b/makewaves/waveio.py
@@ -86,11 +86,8 @@ class WaveGen(QThread):
             daqmx.Val_GroupByChannel,
             rampup_ts,
         )
-
         daqmx.StartTask(self.AOtaskHandle)
-
-        # Wait to allow the DAQmx buffer to empty
-        self.sleep()
+        time.sleep(0.5)
 
         # Set iteration variable to keep track of how many chunks of data
         # have been written
@@ -98,6 +95,7 @@ class WaveGen(QThread):
 
         # Main running loop that writes data to DAQmx buffer
         while self.enable:
+            t0 = time.time()
             print("Writing main wave time series data iteration", i)
             write_space_avail = daqmx.GetWriteSpaceAvail(self.AOtaskHandle)
             print("Write space available:", write_space_avail)
@@ -115,7 +113,7 @@ class WaveGen(QThread):
                 self.dataw,
             )
             i += 1
-            self.sleep()
+            self.sleep(time_taken=time.time() - t0)
 
         print("Output disabled; ramping down")
         # After disabled, initiate rampdown time series
@@ -153,13 +151,16 @@ class WaveGen(QThread):
         daqmx.ClearTask(self.AOtaskHandle)
         self.rampeddown = True
 
-    def sleep(self):
+    def sleep(self, time_taken=0.0):
         """Sleep between iterations."""
-        frac = 0.98
+        frac = 1.0
         if self.wavetype == "Regular":
-            time.sleep(frac * self.period)
+            sleeptime = frac * self.period
         else:
-            time.sleep(frac)
+            sleeptime = frac
+        sleeptime -= time_taken
+        print("Sleeping", sleeptime, "seconds")
+        time.sleep(sleeptime)
 
     def stop(self):
         self.stopgen = WaveStop(self)

--- a/makewaves/waveio.py
+++ b/makewaves/waveio.py
@@ -99,22 +99,22 @@ class WaveGen(QThread):
         # Main running loop that writes data to DAQmx buffer
         while self.enable:
             print("Writing main wave time series data iteration", i)
-            writeSpaceAvail = daqmx.GetWriteSpaceAvail(self.AOtaskHandle)
-            if writeSpaceAvail >= self.buffsize:
-                if self.wavetype != "Regular":
-                    if i >= 120:
-                        self.dataw = tsparts[i % 120, :]
-                    else:
-                        self.dataw = tsparts[i, :]
-                daqmx.WriteAnalogF64(
-                    self.AOtaskHandle,
-                    self.buffsize,
-                    False,
-                    10.0,
-                    daqmx.Val_GroupByChannel,
-                    self.dataw,
-                )
-                i += 1
+            write_space_avail = daqmx.GetWriteSpaceAvail(self.AOtaskHandle)
+            print("Write space available:", write_space_avail)
+            if self.wavetype != "Regular":
+                if i >= 120:
+                    self.dataw = tsparts[i % 120, :]
+                else:
+                    self.dataw = tsparts[i, :]
+            daqmx.WriteAnalogF64(
+                self.AOtaskHandle,
+                self.buffsize,
+                False,
+                10.0,
+                daqmx.Val_GroupByChannel,
+                self.dataw,
+            )
+            i += 1
             self.sleep()
 
         print("Output disabled; ramping down")

--- a/makewaves/waveio.py
+++ b/makewaves/waveio.py
@@ -89,11 +89,8 @@ class WaveGen(QThread):
 
         daqmx.StartTask(self.AOtaskHandle)
 
-        # Wait a second to allow the DAQmx buffer to empty
-        if self.wavetype == "Regular":
-            time.sleep(self.period * 0.99)  # was self.period*0.4
-        else:
-            time.sleep(0.99)
+        # Wait to allow the DAQmx buffer to empty
+        self.sleep()
 
         # Set iteration variable to keep track of how many chunks of data
         # have been written
@@ -158,10 +155,11 @@ class WaveGen(QThread):
 
     def sleep(self):
         """Sleep between iterations."""
+        frac = 0.98
         if self.wavetype == "Regular":
-            time.sleep(0.99 * self.period)
+            time.sleep(frac * self.period)
         else:
-            time.sleep(0.99)  # Was self.period
+            time.sleep(frac)
 
     def stop(self):
         self.stopgen = WaveStop(self)

--- a/makewaves/waveio.py
+++ b/makewaves/waveio.py
@@ -87,7 +87,7 @@ class WaveGen(QThread):
             rampup_ts,
         )
         daqmx.StartTask(self.AOtaskHandle)
-        time.sleep(0.5)
+        self.sleep(time_taken=0.5)
 
         # Set iteration variable to keep track of how many chunks of data
         # have been written
@@ -159,8 +159,9 @@ class WaveGen(QThread):
         else:
             sleeptime = frac
         sleeptime -= time_taken
-        print("Sleeping", sleeptime, "seconds")
-        time.sleep(sleeptime)
+        if sleeptime > 0:
+            print("Sleeping", sleeptime, "seconds")
+            time.sleep(sleeptime)
 
     def stop(self):
         self.stopgen = WaveStop(self)

